src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java

```diff
@@ -156,6 +162,35 @@ public BuilderInfo pollTaskTestingStatus(String taskId) throws TaskNotFoundExcep
     return taskIndex.getTask(taskId).getTestingCopyBuilderInfo();
   }
 
+  @Override
+  public List<String> listFiles(String taskId, Boolean usingTestingVersion, String variant)
+      throws TaskNotFoundException, TaskMissingVariantException, TaskStorageException {
+    Task t = taskIndex.getTask(taskId);
+    LOG.info("List files " + t + ", " + usingTestingVersion + ", " + variant);
+    try (TaskCopy c = usingTestingVersion ? t.acquireTestingCopy() : t.acquireRegisteredCopy()) {
+      if (!c.getInfo().getVariants().contains(variant)) {
+        throw new TaskMissingVariantException("Variant " + variant + " is not defined");
+      }
+
+      return c.listSkeleton(variant);
+    }
+  }
+
+  @Override
+  public Response readFile(String taskId, String fileName, Boolean usingTestingVersion,
+                           String variant)
+      throws TaskNotFoundException, TaskMissingVariantException {
+    Task t = taskIndex.getTask(taskId);
+    try (TaskCopy c = usingTestingVersion ? t.acquireTestingCopy() : t.acquireRegisteredCopy()) {
+      if (!c.getInfo().getVariants().contains(variant)) {
+        throw new TaskMissingVariantException("Variant " + variant + " is not defined");
+      }
+
+      File filePath = new File(c.getSkeletonLocation(variant), fileName);
```

Comment:
don't you need to sanitize filename here to stop people doing things like ../task.json?

Reply:
Oh yeah, we certainly do. Done.